File: LeetCode/77.Combinations/77.Combinations.py
#!/usr/bin/env python3


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:

        res = []
        path = []

        def backtrack(start):
            # print(path)
            if len(path) == k:
                res.append(path[:])
                return

            # The loop upper bound 'n - (k - len(path)) + 2' is a pruning optimization.
            # It ensures there are enough remaining numbers to complete a valid combination,
            # thus reducing unnecessary recursive calls.
            for i in range(start, n - (k - len(path)) + 2):
                path.append(i)
                backtrack(i + 1)
                path.pop()

        backtrack(1)

        return res

File: LeetCode/77.Combinations/77.Combinations.swift
class Solution {
    func combine(_ n: Int, _ k: Int) -> [[Int]] {
        let nums = Array(1 ... n)

        var res: [[Int]] = []
        func backtrack(_ i: Int, _ path: [Int]) {
            if path.count == k {
                res.append(path)
                return
            }

            for j in i ..< n {
                backtrack(j + 1, path + [nums[j]])
            }
        }

        backtrack(0, [])

        return res
    }
}
